refinenet/refinenet.py

```diff
@@ -0,0 +1,269 @@
+        self.data_format = "channels_last"
+
+        tf.reset_default_graph()
+
+        # Set up the placeholder tensors to use as input/output to network
+        image = tf.placeholder(
+            tf.float32, shape=[None, input_shape[0], input_shape[1], 3],
+            name="input_image")
+        annotation = tf.placeholder(
+            tf.int32, shape=[None, input_shape[0], input_shape[1], 1],
+            name="annotation")
+
+        self.placeholders = {"image": image,
+                             "annotation": annotation}
+
+        # Build the computational graph
+        self.predict_with_network, self.logits = self.build_computational_graph()
+
+        self.loss = tf.reduce_mean(
+                       (tf.nn.sparse_softmax_cross_entropy_with_logits(
+                        logits=self.logits, labels=tf.squeeze(annotation, squeeze_dims=[3]), name="entropy")))
+
+        # get train op
+        trainable_var = tf.trainable_variables()
+        self.train_op = self.train(self.loss, trainable_var)
+
+        self.num_params = self.compute_total_num_params()
+
+        self.sess = tf.Session()
+
+        # Set up the saver and try to restore any model previously being trained
+        self.saver = tf.train.Saver()
+
+        # Initialize the graph, then load the network into memory.
+        self.sess.run(tf.global_variables_initializer())
+        if network_storage_dir:
+            self.load_network_weights(self.network_weights_loc)
+
+    def load_network_weights(self, path_to_network):
+        """Load the network weights contained in `path_to_network` into memory."""
+        ckpt = tf.train.get_checkpoint_state(path_to_network)
+        if ckpt and ckpt.model_checkpoint_path:
+            self.saver.restore(self.sess, ckpt.model_checkpoint_path)
+            self.weights_loaded = True
+        else:
+            self.weights_loaded = False
+
+    def compute_total_num_params(self):
+        """Compute the number of parameters of this model"""
+        total_parameters = 0
+        for variable in tf.trainable_variables():
+            shape = variable.get_shape()
+            variable_parameters = 1
+            for dim in shape:
+                variable_parameters *= dim.value
+            total_parameters += variable_parameters
+        return total_parameters
+
+    def build_computational_graph(self):
+        """Build the semantic segmentation network's computational graph."""
+
+        # preprocess the input images by subtracting mean pixel of the batch
+        axis_to_reduce = (0, 1, 2) if self.data_format == "channels_last" else (0, 2, 3)
+        mean_pixel = tf.reduce_mean(self.placeholders["image"], axis=axis_to_reduce)
+        processed_image = self.placeholders["image"] - mean_pixel
+
+        if self.data_format == "channels_first":
+            processed_image = tf.transpose(processed_image, [0, 3, 1, 2])
+
+        num_filters = NUM_REFINE_FILTERS
+
+        def rcu(x):
```

Comment:
Many of the helper functions here are named with acronyms. The code would be easier to read if you used longer names and would remove the need to add comments that explain what the function calls do.